File: provider/src/main/java/com/raygun/raygun4android/messages/crashreporting/RaygunBreadcrumbMessage.java
package com.raygun.raygun4android.messages.crashreporting;

import java.util.Map;
import java.util.WeakHashMap;

public class RaygunBreadcrumbMessage {
    private String message;
    private String category;
    private int level;
    private String type = "Manual";
    private Map<String, Object> customData;
    private Long timestamp = System.currentTimeMillis();
    private String className;
    private String methodName;
    private Integer lineNumber;

    public static class Builder {
        private String message;
        private String category;
        private int level = RaygunBreadcrumbLevel.INFO.ordinal();
        private Map<String, Object> customData = new WeakHashMap<>();
        private String className;
        private String methodName;
        private Integer lineNumber;

        public Builder(String message) {
            this.message = message;
        }

        public Builder category(String category) {
            this.category = category;
            return this;
        }

        public Builder level(RaygunBreadcrumbLevel level) {
            this.level = level.ordinal();
            return this;
        }

        public Builder customData(Map<String, Object> customData) {
            this.customData = customData;
            return this;
        }

        public Builder className(String className) {
            this.className = className;
            return this;
        }

        public Builder methodName(String methodName) {
            this.methodName = methodName;
            return this;
        }

        public Builder lineNumber(Integer lineNumber) {
            this.lineNumber = lineNumber;
            return this;
        }

        public RaygunBreadcrumbMessage build() {
            return new RaygunBreadcrumbMessage(this);
        }
    }

    public String getMessage() {
        return message;
    }

    public String getCategory() {
        return category;
    }

    public RaygunBreadcrumbLevel getLevel() {
        return RaygunBreadcrumbLevel.values()[level];
    }

    public Map<String, Object> getCustomData() {
        return customData;
    }

    public String getClassName() {
        return className;
    }

    public String getMethodName() {
        return methodName;
    }

    public Integer getLineNumber() {
        return lineNumber;
    }

    private RaygunBreadcrumbMessage(Builder builder) {
        message = builder.message;
        category = builder.category;
        level = builder.level;
        customData = builder.customData;
        className = builder.className;
        methodName = builder.methodName;
        lineNumber = builder.lineNumber;
    }
}

File: provider/src/main/java/com/raygun/raygun4android/messages/crashreporting/RaygunBreadcrumbMessage.kt
package com.raygun.raygun4android.messages.crashreporting

import com.google.gson.annotations.SerializedName
import java.util.WeakHashMap

data class RaygunBreadcrumbMessage(
    var message: String?,
    var category: String?,
    @SerializedName("level") private var _level: Int,
    var type: String = "Manual",
    var customData: Map<String, Any?>,
    var timestamp: Long = System.currentTimeMillis(),
    var className: String?,
    var methodName: String?,
    var lineNumber: Int?,
) {
    var level: RaygunBreadcrumbLevel
        get() = RaygunBreadcrumbLevel.entries[_level]
        set(value) {
            _level = value.ordinal
        }

    class Builder(
        internal val message: String?,
    ) {
        private var category: String? = null
        private var level: Int = RaygunBreadcrumbLevel.INFO.ordinal
        private var customData: Map<String, Any?> = WeakHashMap()
        private var className: String? = null
        private var methodName: String? = null
        private var lineNumber: Int? = null

        fun category(category: String?): Builder {
            this.category = category
            return this
        }

        fun level(level: RaygunBreadcrumbLevel): Builder {
            this.level = level.ordinal
            return this
        }

        fun customData(customData: Map<String, Any?>): Builder {
            this.customData = customData
            return this
        }

        fun className(className: String?): Builder {
            this.className = className
            return this
        }

        fun methodName(methodName: String?): Builder {
            this.methodName = methodName
            return this
        }

        fun lineNumber(lineNumber: Int?): Builder {
            this.lineNumber = lineNumber
            return this
        }

        fun build(): RaygunBreadcrumbMessage =
            RaygunBreadcrumbMessage(
                message = message,
                category = category,
                _level = level,
                customData = customData,
                className = className,
                methodName = methodName,
                lineNumber = lineNumber,
            )
    }
}
